server/src/main/java/io/crate/planner/operators/AbstractJoinPlan.java

```diff
@@ -69,13 +86,94 @@ public LogicalPlan rhs() {
 
     @Override
     public List<Symbol> outputs() {
-        if (joinType == JoinType.SEMI) {
-            return lhs.outputs();
-        } else {
-            return Lists.concat(lhs.outputs(), rhs.outputs());
+        return outputs;
+    }
+
+    @Nullable
+    protected AbstractJoinPlan.PrunedOutputsResult pruneOutputs(SequencedCollection<Symbol> outputsToKeep) {
+        LinkedHashSet<Symbol> newOutputsToKeep = new LinkedHashSet<>();
+        LinkedHashSet<Symbol> lhsToKeep = new LinkedHashSet<>();
+        LinkedHashSet<Symbol> rhsToKeep = new LinkedHashSet<>();
+        for (Symbol outputToKeep : outputsToKeep) {
+            SymbolVisitors.intersection(outputToKeep, outputs, newOutputsToKeep::add);
+            SymbolVisitors.intersection(outputToKeep, lhs.outputs(), lhsToKeep::add);
+            SymbolVisitors.intersection(outputToKeep, rhs.outputs(), rhsToKeep::add);
+        }
+        if (joinCondition != null) {
+            SymbolVisitors.intersection(joinCondition, lhs.outputs(), lhsToKeep::add);
+            SymbolVisitors.intersection(joinCondition, rhs.outputs(), rhsToKeep::add);
+        }
+        LogicalPlan newLhs = lhs.pruneOutputsExcept(lhsToKeep);
+        LogicalPlan newRhs = rhs.pruneOutputsExcept(rhsToKeep);
+
+        List<Symbol> newOutputs = List.copyOf(newOutputsToKeep);
+        if (newLhs == lhs && newRhs == rhs && outputs.equals(newOutputs)) {
+            return null;
         }
+
+        return new PrunedOutputsResult(newLhs, newRhs, newOutputs);
     }
 
+    protected record PrunedOutputsResult(LogicalPlan lhs, LogicalPlan rhs, List<Symbol> outputs) {}
+
+    @Nullable
+    protected AbstractJoinPlan.RewriteToFetchResult fetchRewrite(Collection<Symbol> usedColumns) {
```

Comment:
Same here, this adapts rewriteToFetch to also handle outputs and makes it usable for all join sublcasses.